src/levels.rs

```diff
@@ -1,5 +1,1053 @@
+        } else {
+            compact_def.next_range = get_key_range(&compact_def.bot);
+        }
+
+        self.cpt_status
+            .write()
+            .unwrap()
+            .compare_and_add(compact_def)?;
+
+        Ok(())
+    }
+
+    /// Try to fill tables from L0 to be compacted with Lbase. If it can not
+    /// do that, it would try to compact tables from L0 -> L0.
+    fn fill_tables_l0(&self, compact_def: &mut CompactDef) -> Result<()> {
+        if self.fill_tables_l0_to_lbase(compact_def).is_err() {
+            return self.fill_tables_l0_to_l0(compact_def);
+        }
+        Ok(())
+    }
+
+    fn fill_tables(&self, compact_def: &mut CompactDef) -> Result<()> {
+        let this_level = compact_def.this_level.read().unwrap();
+        let next_level = compact_def.next_level.read().unwrap();
+
+        let tables = &this_level.tables;
+
+        if tables.is_empty() {
+            return Err(Error::CustomError("no tables to compact".to_string()));
+        }
+
+        // TODO: Sort tables by heuristic.
+
+        // TODO: Don't hold cpt_status write lock for long time.
+        let mut cpt_status = self.cpt_status.write().unwrap();
+
+        for table in tables {
+            compact_def.this_size = table.size();
+            compact_def.this_range = get_key_range_single(table);
+            // If we're already compacting this range, don't do anything.
+            // TODO: ?
+            if cpt_status.overlaps_with(compact_def.this_level_id, &compact_def.this_range) {
+                continue;
+            }
+            compact_def.top = vec![table.clone()];
+            let (left, right) = next_level.overlapping_tables(&compact_def.this_range);
+
+            // TODO: Try to fix it.
+            if right < left {
+                eprintln!(
+                    "right {} is less than left {} in overlapping_tables for current level {}, next level {}, key_range {:?}",
+                    right,
+                    left,
+                    compact_def.this_level_id,
+                    compact_def.next_level_id,
+                    compact_def.this_range
+                );
+                continue;
+            }
+
+            compact_def.bot = next_level.tables[left..right].to_vec();
+
+            if compact_def.bot.is_empty() {
+                compact_def.bot = vec![];
+                compact_def.next_range = compact_def.this_range.clone();
+                if cpt_status.compare_and_add(compact_def).is_err() {
+                    continue;
+                }
+                return Ok(());
+            }
+
+            compact_def.next_range = get_key_range(&compact_def.bot);
+
+            if cpt_status.overlaps_with(compact_def.next_level_id, &compact_def.next_range) {
+                continue;
+            }
+
+            if cpt_status.compare_and_add(compact_def).is_err() {
+                continue;
+            }
+
+            return Ok(());
```

Comment:
Seems it will only fill one compaction job at a time? Is it expected?

Reply:
Yes, for one specified compactor, it will compact for every level serially.